pkg/metadata/manager.go

```diff
@@ -0,0 +1,404 @@
+package metadata
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cilium/cilium/pkg/cgroups"
+	"github.com/cilium/cilium/pkg/defaults"
+	v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodetypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "metadata")
+	// example cgroup path in kubernetes environments
+	// /run/cilium/cgroupv2/kubepods/burstable/pod1858680e-b044-4fd5-9dd4-f137e30e2180/e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0
+	defaultCgroupBasePath = cgroups.GetCgroupRoot() + "/kubepods"
+	// example cgroup path in nested kubernetes environments like kind
+	// /run/cilium/cgroupv2/kubelet/kubepods/pod4841248b-fc2f-41f4-9981-a685bf840ab5/d8f227cc24940cfdce8d8e601f3b92242ac9661b0e83f0ea57fdea1cb6bc93ec
+	defaultNestedCgroupBasePath = cgroups.GetCgroupRoot() + "/kubelet" + "/kubepods"
+	cgroupBasePaths             = []string{defaultCgroupBasePath, defaultNestedCgroupBasePath}
+)
+
+// Manager maintains Kubernetes and low-level metadata (cgroup path and
+// cgroup id) for local pods and their containers. In order to do that, it defines
+// and implements callback functions that are called on Kubernetes pod watcher events.
+// It also exposes APIs to read the saved metadata.
+//
+// During initialization, the manager checks for a valid base cgroup path.
+// In case of environments using non-default cgroup base paths, manager will
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type Manager struct {
+	// Map of pod metadata indexed by their ids
+	podMetadataById map[string]*podMetadata
+	// Map of container metadata indexed by their cgroup ids
+	containerMetadataByCgrpId map[uint64]*containerMetadata
+	// Set to the valid cgroup base path if found
+	templateCgroupBasePath string
+	// Buffered channel to receive pod events
+	podEvents chan podEvent
+	// Object to check cgroup base path
+	checkCgroupPath *sync.Once
+	// Flag to check if manager is enabled, and processing events
+	enabled atomic.Bool
```

Comment:
Why does enabled need to be atomic? It looks like it's always set true from the caller of NewManager and then that goroutine starts the other goroutine which can check the bool.

Did you plan to make this able to be dynamically enabled/disabled?

Reply:
So the flag is enabled depending on whether socket-lb is enabled or not. I didn't initially realize that I had added the flag in the list that's enabled dynamically at run time. Anyway the latest push reflects the right intentions. I think this also addresses your above comment.